README.md
# Fastify Method Not Allowed

[![js-semistandard-style](https://img.shields.io/badge/code%20style-semistandard-brightgreen.svg)](https://github.com/standard/semistandard)
[![Build Status](https://travis-ci.com/gkampitakis/fastify-method-not-allowed.svg?branch=master)](https://travis-ci.com/gkampitakis/fastify-method-not-allowed)

`fastify-method-not-allowed` is a plugin for returning **405** status code for routes instead of default **404**.

Note you have to register this plugin before registering any routes, as it uses [onRoute](https://www.fastify.io/docs/v3.7.x/Hooks/#onroute) hook, so needs to "collect" them.

## Install

```bash
npm install fastify-method-not-allowed
```

## Usage

Require the module and just register it as any other fastify plugin.

```javascript
const methodNotAllowed = require('fastify-method-not-allowed');

server.register(methodNotAllowed);

server.get('/', (req, res) => res.send('Hello World'));

fastify.listen(3000);

/*
curl -X POST localhost:3000

This will return statusCode 405 instead of the 404 default response
*/
```

## API

### register plugin

```javascript
const methodNotAllowed = require('fastify-method-not-allowed');

server.register(methodNotAllowed, {
    responses: {
        404: 'Not Found',
        405: 'Method Not Allowed',
    },
});
```

### register options

-   `prefix`: prefix on which path this plugin will be enabled
    <br> <span style="font-size:15px"> You can register the plugin multiple times with different prefixes if you want to apply different responses on different paths<span>
-   `responses`
    <br> <span style="font-size:15px"> Default values<span>

    -   404

    ```json
    {
        "message": "Route <method>:<url> not found",
        "error": "Not Found",
        "statusCode": 404
    }
    ```

    -   405

    ```json
    {
        "message": "Route <method>:<url> not allowed",
        "error": "Method Not Allowed",
        "statusCode": 405
    }
    ```

    You can override those values by providing either a **value** or a **function**.

    -   For **404** you have access to the `request` object so you can pass a function e.g. `(req)=>"Route `\${req.url}` not found"`.
    -   For **405** you have access to the `request` and `allowedMethods` so you can pass a function e.g. `(req,methods)=>"Route `\${req.url}` allowed only ${methods}"`.

-   `setNotFoundHandlerOptions` You can pass options to fastify handler. [For more information](https://www.fastify.io/docs/latest/Server/#setnotfoundhandler).

## Typescript

In order to use this plugin you need to enable the flag `"esModuleInterop": true` in `tsconfig.json`.

then you can import it

```typescript
import methodNotAllowed from 'fastify-method-not-allowed';
```

### Note:

This plugin uses the `setNotFoundHandler` handler so you won't be able to re-use this handler in your own service at least with the same `prefix`.

### Changelog

[CHANGELOG.md](./CHANGELOG.md)

### Issues

For any [issues](https://github.com/gkampitakis/fastify-method-not-allowed/issues).

## License

MIT License

CHANGELOG.md
## [1.0.2](https://github.com/gkampitakis/fastify-method-not-allowed/compare/v1.0.1...v1.0.2) (2020-11-15)


### Bug Fixes

* drop unused dev deps ([1a53797](https://github.com/gkampitakis/fastify-method-not-allowed/commit/1a53797ef8a6ee311ab7c0f4f58c36c8c25d9fad))

## [1.0.1](https://github.com/gkampitakis/fastify-method-not-allowed/compare/v1.0.0...v1.0.1) (2020-11-15)


### Bug Fixes

* wrong description ([88bb6e8](https://github.com/gkampitakis/fastify-method-not-allowed/commit/88bb6e8b7c8b309bb6168c4f47095cd5515051c9))

# 1.0.0 (2020-11-15)


### Bug Fixes

* add types, readme documentation and travis pipeline ([7f77e18](https://github.com/gkampitakis/fastify-method-not-allowed/commit/7f77e180f61fc2025358275b98262aaa91029c7a))


lib/tests/methodNotAllowed.spec.js
'use strict';

const Fastify = require('fastify');
const t = require('tap');
const test = t.test;
const methodNotAllowed = require('../');

const handler = (_, __) => { };

function setupRoutes (fastify) {
  fastify.register((instance, _, done) => {
    instance.get('/', handler);
    instance.route({ url: '/', method: ['PATCH', 'OPTIONS'], handler });
    done();
  }, { prefix: '/mock' });// GET, PATCH, OPTIONS

  fastify.register((instance, _, done) => {
    instance.get('/', handler);// GET
    done();
  }, { prefix: '/test' });

  fastify.get('/', handler); // GET
}

test('default options', t => {
  const fastify = Fastify();

  fastify.register(methodNotAllowed);
  setupRoutes(fastify);

  test('it should return 405', t => {
    t.plan(6);

    fastify.inject({ url: '/', method: 'POST' }, (err, res) => {
      const body = JSON.parse(res.body);

      t.is(err, null);
      t.is(res.statusCode, 405);
      t.is(body.statusCode, 405);
      t.is(res.headers.allow, 'GET');
      t.is(body.message, 'Route POST:/ not allowed');
      t.is(body.error, 'Method Not Allowed');
    });
  });

  test('it should return 404', t => {
    t.plan(5);

    fastify.inject({ url: '/not-found', method: 'POST' }, (err, res) => {
      const body = JSON.parse(res.body);

      t.is(err, null);
      t.is(res.statusCode, 404);
      t.is(body.statusCode, 404);
      t.is(body.message, 'Route POST:/not-found not found');
      t.is(body.error, 'Not Found');
    });
  });

  t.end();
});

test('handle only prefixed routes', t => {
  const fastify = Fastify();

  fastify.register(methodNotAllowed, { prefix: '/mock' });
  fastify.register(methodNotAllowed, { prefix: '/test' });
  setupRoutes(fastify);

  test('non-prefixed route should return 404', t => {
    t.plan(5);

    fastify.inject({ url: '/', method: 'POST' }, (err, res) => {
      const body = JSON.parse(res.body);

      t.is(err, null);
      t.is(res.statusCode, 404);
      t.is(body.statusCode, 404);
      t.is(body.message, 'Route POST:/ not found');
      t.is(body.error, 'Not Found');
    });
  });

  test('prefixed route should return 405', t => {
    t.plan(6);

    fastify.inject({ url: '/mock', method: 'POST' }, (err, res) => {
      const body = JSON.parse(res.body);

      t.is(err, null);
      t.is(res.statusCode, 405);
      t.is(body.statusCode, 405);
      t.is(res.headers.allow, 'GET, PATCH, OPTIONS');
      t.is(body.message, 'Route POST:/mock not allowed');
      t.is(body.error, 'Method Not Allowed');
    });
  });

  test('prefixed route should return 405', t => {
    t.plan(6);

    fastify.inject({ url: '/test', method: 'POST' }, (err, res) => {
      const body = JSON.parse(res.body);

      t.is(err, null);
      t.is(res.statusCode, 405);
      t.is(body.statusCode, 405);
      t.is(res.headers.allow, 'GET');
      t.is(body.message, 'Route POST:/test not allowed');
      t.is(body.error, 'Method Not Allowed');
    });
  });

  t.end();
});

test('pass different response message', t => {
  const fastify = Fastify();

  fastify.register(methodNotAllowed, {
    prefix: '/mock',
    responses: {
      405: (request, allowedMethods) => ({
        message: `Custom message on ${request.method}:${request.url}`,
        allowed: allowedMethods
      }),
      404: 'Route not found'
    }
  });
  fastify.register(methodNotAllowed, { prefix: '/test' });
  setupRoutes(fastify);

  test('non-prefixed route should return 404', t => {
    t.plan(5);

    fastify.inject({ url: '/', method: 'POST' }, (err, res) => {
      const body = JSON.parse(res.body);

      t.is(err, null);
      t.is(res.statusCode, 404);
      t.is(body.statusCode, 404);
      t.is(body.message, 'Route POST:/ not found');
      t.is(body.error, 'Not Found');
    });
  });

  test('prefixed route should return custom 405 response', t => {
    t.plan(7);

    fastify.inject({ url: '/mock', method: 'POST' }, (err, res) => {
      const body = JSON.parse(res.body);

      console.log(body.allowed);

      t.is(err, null);
      t.is(res.statusCode, 405);
      t.is(body.statusCode, undefined);
      t.is(body.message, 'Custom message on POST:/mock');
      t.same(body.allowed, ['GET', 'PATCH', 'OPTIONS']);
      t.is(res.headers.allow, 'GET, PATCH, OPTIONS');
      t.is(body.error, undefined);
    });
  });

  test('prefixed route should return custom 404 response', t => {
    t.plan(3);

    fastify.inject({ url: '/mock/not-found', method: 'POST' }, (err, res) => {
      t.is(err, null);
      t.is(res.statusCode, 404);
      t.is(res.body, 'Route not found');
    });
  });

  test('prefixed route should return default 405', t => {
    t.plan(6);

    fastify.inject({ url: '/test', method: 'POST' }, (err, res) => {
      const body = JSON.parse(res.body);

      t.is(err, null);
      t.is(res.statusCode, 405);
      t.is(body.statusCode, 405);
      t.is(body.message, 'Route POST:/test not allowed');
      t.is(res.headers.allow, 'GET');
      t.is(body.error, 'Method Not Allowed');
    });
  });

  test('prefixed route should return default 404', t => {
    t.plan(5);

    fastify.inject({ url: '/test', method: 'POST' }, (err, res) => {
      const body = JSON.parse(res.body);

      t.is(err, null);
      t.is(res.statusCode, 405);
      t.is(body.statusCode, 405);
      t.is(body.message, 'Route POST:/test not allowed');
      t.is(body.error, 'Method Not Allowed');
    });
  });

  t.end();
});

test('pass setNotFoundHandler options', t => {
  const fastify = Fastify();

  fastify.register(methodNotAllowed, {
    setNotFoundHandlerOptions: {
      preHandler: (req, reply, done) => {
        reply.send('message');
        done();
      }
    }
  });
  setupRoutes(fastify);

  test('setNotFoundHandlerOptions should by applied', t => {
    t.plan(3);

    fastify.inject({ url: '/', method: 'POST' }, (err, res) => {
      t.is(err, null);
      t.is(res.statusCode, 200);
      t.is(res.body, 'message');
    });
  });

  t.end();
});


lib/src/methodNotAllowed.js
'use strict';

const { PathnameStore } = require('@lunjs/pathname-store');
const defaultResponses = {
  404: ({ method, url }) => ({
    message: `Route ${method}:${url} not found`,
    error: 'Not Found',
    statusCode: 404
  }),
  405: ({ method, url }) => ({
    message: `Route ${method}:${url} not allowed`,
    error: 'Method Not Allowed',
    statusCode: 405
  })
};

function registerRoutes (fastify, routeRegistry, options) {
  fastify.addHook('onRoute', (routeOptions) => {
    const { url, prefix, method } = routeOptions;

    if (options.prefix && options.prefix !== prefix) return;

    const _methods = Array.isArray(method) ? method : [method];
    const item = routeRegistry.find(url);

    if (item.found) {
      routeRegistry.add(url, item.box.store.concat(_methods));
      return;
    }

    routeRegistry.add(url, _methods);
  });
}

function returnCallable (unknownObj) {
  if (typeof (unknownObj) === 'function') return unknownObj;
  else return () => unknownObj;
}

function methodNotAllowed (fastify, options, next) {
  const routeRegistry = new PathnameStore();
  registerRoutes(fastify, routeRegistry, options);

  const { setNotFoundHandlerOptions, prefix, responses = {} } = options;
  const computedResponses = {
    404: responses[404] ? returnCallable(responses[404]) : defaultResponses[404],
    405: responses[405] ? returnCallable(responses[405]) : defaultResponses[405]
  };

  fastify.register(function (instance, _, done) {
    instance.setNotFoundHandler(setNotFoundHandlerOptions, (req, res) => {
      const value = routeRegistry.find(req.url.split('?')[0]);
      if (value.found) {
        return res
          .status(405)
          .header('Allow', value.box.store.join(', '))
          .send(computedResponses[405](req, value.box.store));
      }

      return res.status(404).send(computedResponses[404](req));
    });

    done();
  }, { prefix });

  next();
}

module.exports = {
  methodNotAllowed
};
